File: 09-des/09-des.py
# --- Day 9: Rope Bridge ---

# Tidlig-stadie av traumatisk kode og forsøk på debugging i python

def adjacent_nodes(node1, node2):
    return -1 <= node1[0] - node2[0] <= 1 and -1 <= node1[1] - node2[1] <= 1


def get_unique_tail_positions(head, tail, moves):
    tail_positions = []

    for move in moves:
        for _ in range(move[1]):
            if move[0] == 'R':
                head[0] += 1
                if not adjacent_nodes(tail, head):
                    tail[0] = head[0]-1
                    tail[1] = head[1]
            elif move[0] == 'L':
                head[0] -= 1
                if not adjacent_nodes(tail, head):
                    tail[0] = head[0]+1
                    tail[1] = head[1]
            elif move[0] == 'U':
                head[1] += 1
                if not adjacent_nodes(tail, head):
                    tail[0] = head[0]
                    tail[1] = head[1]-1
            elif move[0] == 'D':
                head[1] -= 1
                if not adjacent_nodes(tail, head):
                    tail[0] = head[0]
                    tail[1] = head[1]+1
            tail_positions.append(tuple(tail))

    return len(set(tail_positions))


def main():
    with open("09-des.txt", "r") as file:
        moves = []
        for line in file:
            # line = line.split()
            moves.append([line[0], int(line[1])])

    head = [0, 0]
    tail = [0, 0]

    # print(moves)

    print("Part 1: ", get_unique_tail_positions(head, tail, moves))


if __name__ == "__main__":
    main()

File: 09-des/09-des.jl
# --- Day 9: Rope Bridge ---

struct Move
    dir::Char
    dist::Int
end

function Move(moves::Vector{Char})
    return Move(moves[1], parse(Int, join(moves[2:end])))
end

function get_positions(moves)
    node = [0, 0]
    positions = []

    for move in moves
        for _ in 1:move.dist
            if move.dir == 'R'
                node[1] += 1
            elseif move.dir == 'L'
                node[1] -= 1
            elseif move.dir == 'U'
                node[2] += 1
            elseif move.dir == 'D'
                node[2] -= 1
            end
            push!(positions, Tuple(node))
        end
    end
    return positions
end

function follow_node(moves)
    node = [0, 0]
    positions = []

    for move in moves
        if abs(move[1] - node[1]) > 1 || abs(move[2] - node[2]) > 1
            node[1] += (move[1] > node[1]) - (move[1] < node[1])
            node[2] += (move[2] > node[2]) - (move[2] < node[2])
        end
        push!(positions, Tuple(node))
    end
    return positions
end

function main()
    moves = Move.(deleteat!.(collect.(readlines("09-des.txt")), 2))

    ### Note to self: Husk å sjekke at du har lest input riktig. 
    ### Da slipper du å prøve å debugge i flere timer uten å skjønne hva som er feil...

    p1 = p2 = get_positions(moves)
    for _ in 1:9
        p2 = follow_node(p2)
    end
    
    println("Part 1: ", length(Set(follow_node(p1))))
    println("Part 2: ", length(Set(p2)))

end

main()
